cpp/src/io/csv/csv_gpu.cu

```diff
@@ -1014,23 +1016,30 @@ void __host__ remove_blank_rows(rmm::device_vector<uint64_t> &row_offsets,
   row_offsets.resize(new_end - row_offsets.begin());
 }
 
-cudaError_t __host__ DetectColumnTypes(const char *data,
-                                       const uint64_t *row_starts,
-                                       size_t num_rows,
-                                       size_t num_columns,
-                                       const ParseOptions &options,
-                                       column_parse::flags *flags,
-                                       column_parse::stats *stats,
-                                       cudaStream_t stream)
+std::vector<column_parse::stats> detect_column_types(const char *data,
+                                                     const uint64_t *row_starts,
+                                                     size_t num_rows,
+                                                     size_t num_actual_columns,
+                                                     size_t num_active_columns,
+                                                     const cudf::io::ParseOptions &options,
+                                                     column_parse::flags *flags,
+                                                     cudaStream_t stream)
 {
   // Calculate actual block count to use based on records count
   const int block_size = csvparse_block_dim;
   const int grid_size  = (num_rows + block_size - 1) / block_size;
 
+  auto d_stats = rmm::device_vector<column_parse::stats>(num_active_columns);
+
   data_type_detection<<<grid_size, block_size, 0, stream>>>(
-    data, options, num_rows, num_columns, flags, row_starts, stats);
+    data, options, num_rows, num_actual_columns, flags, row_starts, d_stats.data().get());
 
-  return cudaSuccess;
+  auto h_stats = std::vector<column_parse::stats>();
+  h_stats.reserve(num_active_columns);
+
+  thrust::copy(d_stats.begin(), d_stats.end(), h_stats.begin());
+
+  return h_stats;
```

Comment:
If `thrust::copy` works the the same as `std::copy` there is a bug here. `reserve` only sets `capacity`, not the `size`: https://www.godbolt.org/z/jG46Yz

Reply:
Yeah, I was wondering about that, but it seems to work? Still not great for readability. I'll just initialize it, since there's no way around it for an `std::vector` anyway (afaik). :shrug:

Reply:
Can `thrust::copy` work with a `back_inserter`? AFAIK that's the only way to avoid double init (at least with `std::copy`)

Reply:
Why not use a range constructor (aka the #5 on https://en.cppreference.com/w/cpp/container/vector/vector). Code ends up with a single line:
```cpp
  auto h_stats = std::vector<column_parse::stats>();
  h_stats.reserve(num_active_columns);

  thrust::copy(d_stats.begin(), d_stats.end(), h_stats.begin());

  return h_stats;
```
becomes:
```cpp
  return std::vector<column_parse::stats>(d_stats.begin(), d_stats.end());
```

Reply:
@vuule Makes sense to me. I'll give that a shot.

Reply:
`std::vector` does not know how to copy from a device pointer.  You would need to use `thrust::host_vector`.